Primer entregable

FernandoGal25 opened the pull request:
-Se crea modelo de dominio con nombre generico
-Se crea caso de uso para obtener los elementos del csv
-Se crea manejador de archivos csv
-Se genera un inyector de dependencias
-Se crea repositorio para busqueda por id
-Se genera ruta GET para la obtencion de items
-Se crea controlador padre que encapsula controlador hijo encargado de iniciar el caso de uso
-Se genera una interfaz de contexto para abstraer la logica del contexto del framework Echo

Commit:
Endpoint de lectura de archivo CSV
-Se crea modelo de dominio con nombre generico
-Se crea caso de uso para obtener los elementos del csv
-Se crea manejador de archivos csv
-Se genera un inyector de dependencias
-Se crea repositorio para busqueda por id
-Se genera ruta GET para la obtencion de items
-Se crea controlador padre que encapsula controlador hijo encargado de iniciar el caso de uso
-Se genera una interfaz de contexto para abstraer la logica del contexto del framework Echo

RolloCasanova commented on application/contracts/item_repository.go:
```diff
@@ -0,0 +1,8 @@
+
+type ItemRepository interface {
+	FindByID(id uint64) (*model.Item, error)
+	// Create(u *model.Item) (*model.Item, error)
```
Remove unnecessary comment.

RolloCasanova commented on application/usecase/item.go:
```diff
@@ -0,0 +1,38 @@
+
+type ItemService interface {
+	GetByKey(key uint64) (*model.Item, error)
+	// Create(u *model.Item) (*model.Item, error)
```
Remove unnecessary comment.

RolloCasanova commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+
+	"github.com/davecgh/go-spew/spew"
+
+	"github.com/spf13/viper"
```
```suggestion
import (
	"fmt"
	"log"
	"os"
	"path/filepath"

	"github.com/davecgh/go-spew/spew"
	"github.com/spf13/viper"
```

RolloCasanova commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+
+	viper.SetConfigName("config")
+	viper.SetConfigType("yml")
+	path, _ := os.Getwd()
```
Never ignore errors.

RolloCasanova commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+	viper.AutomaticEnv()
+
+	if err := viper.ReadInConfig(); err != nil {
+		fmt.Println(err)
```
No need to use `fmt.Println()` method. You are already logging the error.

Also, as a previous comment in this function, it would be better here to return nil and the error.

RolloCasanova commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+	}
+
+	if err := viper.Unmarshal(&Config); err != nil {
+		fmt.Println(err)
```
You should use log instead of fmt. fmt is used for debugging purposes, or small/simple console programs, not intended to be used on bigger projects.

Also, as a previous comment in this function, it would be better here to return nil and the error.

RolloCasanova commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+	}
+}
+
+var C config
```
Let's try to avoid having global variables. In this case, this is an input for your function that is not explicitly defined as so; this might lead to painful testing in a future.

RolloCasanova commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+
+var C config
+
+func ReadConfig() {
```
I'd suggest changing the function signature to return a config pointer and an error; in the place where you are using it, handle the error (if that's the case), or simply use the obtained config.
```suggestion
func ReadConfig() (*config, error) {
```

RolloCasanova commented on config/config.go:
```diff
@@ -0,0 +1,45 @@
+
+	if err := viper.Unmarshal(&Config); err != nil {
+		fmt.Println(err)
+		os.Exit(1)
```
Do not handle Exit here; this should be done in the main function.

RolloCasanova commented on application/usecase/item.go:
```diff
@@ -0,0 +1,38 @@
+	// Create(u *model.Item) (*model.Item, error)
+}
+
+func NewItemService(r contracts.ItemRepository) ItemService {
```
There's a saying in go: `Accept Interfaces, Return Concrete Types`.

Please, fix this and all other occurrences of this issue.

RolloCasanova commented on interface/controller/item_controller.go:
```diff
@@ -0,0 +1,54 @@
+
+type ItemController interface {
+	GetItemById(c Context) error
+	// CreateItem(c Context) error
```
Remove unnecessary comment.

RolloCasanova commented on interface/controller/item_controller.go:
```diff
@@ -0,0 +1,54 @@
+	return &itemController{us}
+}
+
+func (uc *itemController) GetItemById(c Context) error {
```
For better readability, acronyms should be declared in all uppercases.
```suggestion
func (uc *itemController) GetItemByID(c Context) error {
```

RolloCasanova commented on interface/controller/item_controller.go:
```diff
@@ -0,0 +1,54 @@
+		return err2
+	}
+	return c.JSON(http.StatusOK, item)
+}
```
Normally, err assignation and checks are performed one after the other with no empty line.

Also, you should declare variables as soon as you are going to use them; if there's an error in line 27, there was no sense having created the `item` var in line 26. 

Even better, there's no need to declare it beforehand; you could use the short variable declaration. Go will infer its type, and this way we could also reuse `err`, instead of creating new errors here and there.

```suggestion
	key, err := strconv.ParseUint(c.Param("id"), 10, 16)
	if err != nil {
		return errors.New("QueryParam invalido")
	}
	
	item, err := uc.Usecase.GetByKey(key)
	if err != nil {
		return err
	}
	
	return c.JSON(http.StatusOK, item)
}
```

RolloCasanova commented on interface/controller/item_controller.go:
```diff
@@ -0,0 +1,54 @@
+	var item *model.Item
+	key, err := strconv.ParseUint(c.Param("id"), 10, 16)
+	if err != nil {
+		return errors.New("QueryParam invalido")
```
You should wrap errors from other packages. This will help you in the future when debugging or tracking the root cause of an issue.

Please fix this and all other occurrences.

RolloCasanova replied:
When the conversion of query param `id` fails it just returns an Internal Server Error message. You could create your own project errors definition and display a little more information to let the user know what's going on.

RolloCasanova commented on interface/controller/item_controller.go:
```diff
@@ -0,0 +1,54 @@
+
+	return c.JSON(http.StatusCreated, u)
+}
+*/
```
Please delete unnecessary comments. Instead of having them commented, better put them in another branch.

RolloCasanova commented on interface/repository/csv_item_repository.go:
```diff
@@ -0,0 +1,62 @@
+	"errors"
+	"io"
+	"strconv"
+)
```
Imported packages should be organized into three groups:
1. Golang's libraries
2. Internal packages
3. Third-party packages
```suggestion
import (
	"errors"
	"io"
	"strconv"
	
	"academy_bootcamp/application/contracts"
	"academy_bootcamp/domain/model"
	"academy_bootcamp/infrastructure/datastore"
)
```

RolloCasanova commented on interface/repository/csv_item_repository.go:
```diff
@@ -0,0 +1,62 @@
+		record, err := r.Handler.Read()
+		if err == io.EOF {
+			break
+		}
```
Why name it `erro` instead of the well-known `err` name?

No need to leave blank lines after the for.

Cuddle the assignation and checking of your errors.

```suggestion
func (r *CSVItemRepository) FindByID(id uint64) (*model.Item, error) {
	var rec *model.Item

	err := r.Handler.BuildHandler()
	if err != nil {
		return nil, err
	}
	for {
		record, err := r.Handler.Read()
		if err == io.EOF {
			break
		}
```

RolloCasanova commented on interface/repository/csv_item_repository.go:
```diff
@@ -0,0 +1,62 @@
+
+		if err2 != nil {
+			return nil, errors.New("no es una llave")
+		}
```
Cuddle the assignation and checking of your errors.

No need to declare more errors; you could re-use `err`.
```suggestion
		key, err2 := strconv.ParseUint(record[0], 10, 16)
		if err2 != nil {
			return nil, errors.New("no es una llave")
		}
```

RolloCasanova commented on interface/repository/csv_item_repository.go:
```diff
@@ -0,0 +1,62 @@
+	}
+
+	return u, nil
+}*/
```
Delete these lines.

RolloCasanova commented on main.go:
```diff
@@ -0,0 +1,29 @@
+	e := echo.New()
+	e = router.NewRouter(e, r.Register())
+
+	fmt.Println("Server listen at http://localhost" + ":" + config.C.Server.Port)
```
Use log instead of fmt.

RolloCasanova commented on application/usecase/item.go:
```diff
@@ -0,0 +1,38 @@
+func (s *itemService) GetByKey(key uint64) (*model.Item, error) {
+	u, err := s.ItemRepository.FindByID(key)
+	if err != nil {
+		return nil, err
```
Should wrap errors from external packages.

RolloCasanova commented on application/usecase/item.go:
```diff
@@ -0,0 +1,38 @@
+	}
+
+	return u, nil
+}*/
```
Delete these lines.

RolloCasanova commented on infrastructure/datastore/csv.go:
```diff
@@ -0,0 +1,39 @@
+func (h *CSVHandler) BuildHandler() error {
+	f, err := os.Open(h.Path)
+	if err != nil {
+		return errors.New("no se pudo abrir el archivo")
```
Do not mix English with other languages.

Also, it's better to declare and use errors instead of dynamically creating them on the fly.

RolloCasanova commented on application/usecase/item.go:
```diff
@@ -0,0 +1,38 @@
+	return &itemService{r}
+}
+
+func (s *itemService) GetByKey(key uint64) (*model.Item, error) {
```
All exported elements (funcs, vars, consts, types, etc.) should contain a comment in the line above its definition with a brief description of what it does, and what it can return.

This is the way you could document your code; give it a look to any Golang internal library/package to see what I'm talking about.

Please, let's try to follow best practices and add these comments in exported elements.

RolloCasanova commented on go.mod:
```diff
@@ -0,0 +1,38 @@
+module academy_bootcamp
```
Your module should be named the same way you have it stored in a code repository (github, gitlab, etc.); you should follow the same path.

In this case, should be `github.com/FernandoGal25/academy-go-q42021`; this is done this way because other people might import your package, and avoid conflicts if another package is exactly named as you have it right now (not probable).

RolloCasanova commented on domain/model/item.go:
```diff
@@ -0,0 +1,7 @@
+package model
+
+// Se deja el nombre generico item en lo que se encuentra un correcto nombre para el modelo de negocio
```
Remember the format on how to document your exported items; let's try to follow these best practices.

Also, IMHO, I don't like the idea of mixing languages (English and Spanish); even these are comments, people who don't know Spanish will be struggling to read them.

RolloCasanova commented on infrastructure/router/router.go:
```diff
@@ -0,0 +1,18 @@
+	e.Use(middleware.Recover())
+
+	e.GET("/items/:id", func(context echo.Context) error { return c.Item.GetItemById(context) })
+	// e.POST("/items", func(context echo.Context) error { return c.Item.CreateItem(context) })
```
Delete this unnecessary comment.

RolloCasanova commented on interface/controller/item_controller.go:
```diff
@@ -0,0 +1,54 @@
+	if err2 != nil {
+		return err2
+	}
+	return c.JSON(http.StatusOK, item)
```
`JSON()` method is returning an error; you should always handle errors.

RolloCasanova commented on infrastructure/router/router.go:
```diff
@@ -0,0 +1,18 @@
+	e.Use(middleware.Logger())
+	e.Use(middleware.Recover())
+
+	e.GET("/items/:id", func(context echo.Context) error { return c.Item.GetItemById(context) })
```
You should never ignore errors; remember to also wrap and propper handle it.

RolloCasanova commented on public/csv/file.csv:
```diff
@@ -0,0 +1,3 @@
+1,bulbasaur
```
Please add a little more info in the file.

RolloCasanova commented on interface/controller/item_controller.go:
```diff
@@ -0,0 +1,54 @@
+		return errors.New("QueryParam invalido")
+	}
+
+	item, err2 := uc.Usecase.GetByKey(key)
```
If you make a request with an inexisting id, it is returning a `null` and a `200 OK` response. Please fix this.

RolloCasanova reviewed (changes requested):
Please, move your `README.md` file content to another file, and in this file add information regarding your project; a brief description, requirements, what it does, which endpoints it handles, how to run it, some examples, some TODO section, etc. so users could know in a glance what is this about.

Also, it could be a good idea to add an endpoint to get all items.

Please also try to avoid writing in Spanish (comments, code, and even the commit messages).

I liked your approach for clean architecture, and liked it more that you're using interfaces; this will be helpful to create unit tests and make your project easier to scale up.

Any doubt about any comment please reply to it, or send me a DM to talk about it.

Good job!

Commit:
Fixes of GET endpoint
- Creates new GET endpoint that return all items
- Specifies domain entity being Pokemon
- Wraps errors and handles each to generate readable responses
- Addes code comments
- Generates unit tests for each endpoint
- Creates a pokemon mock repository that helps in the unit testing
